### PR TITLE
WIP -- Add a storage backend that works for DSE-5.0

### PR DESCRIPTION
**Pull request by michaelsembwever**
Add a storage backend that works for DSE-5.0
  This would be helpful for those that can't upgrade to cassandra-3.9

### CODE REVIEW

**Comment by michaelsembwever** [comment]
Basic thoughts and a skeleton for using Zipkin with the new style of Cassandra datamodel, but replacing the SASI with Solr Cores. This allows the data model to run against Cassandra-3.0.8+ but does require a DSE installation.

A fair bit of work still has to happen here, wiring it together, etc.
But the basic idea is in place.

**Comment by codefromthecrypt** [comment]
interested, as I'm interested in decommissioning the "cassandra" module, but I don't think this gets us there, as not everyone uses DSE. Basically, we end up with 3 codebases for cassandra that don't share code.

I know @schrepfler asked about DSE, but not sure if he uses zipkin at the moment or not. Do we have a good gauge of how many use DSE?

I'm concerned about the maintenance problem of bringing this into the core repo, particularly as it is getting bloated and it doesn't seem obvious how to test commercial variants like this. There's also bus factor.. we're lucky we have you, but there's few others working on cassandra, almost none. Can we afford to have a third codebase? Finally, we already blow up on memory often as it is and our builds are approaching 20minutes.. Feels like it should be a separate repo.

@openzipkin/cassandra @abesto What are your thoughts?

**Comment by michaelsembwever** [comment]
Personally I would rather not include the code into master, for the same reasons you describe @adriancole 

But I did want to at least throw the code out here, so that others could take it (if only the idea) and run if they wanted to.

**Comment by codefromthecrypt** [comment]
@michaelsembwever I noticed this has a dep on cassandra3. seems `CassandraSpanConsumer` is the same type, but it has a different front-end (SpanStore) and schema.

Am I right to think that this is "effectively" the same schema? Else, I'd assume you wouldn't be able to use cassandra3's `CassandraSpanConsumer`

While this PR is >2K lines, it seems some of it is benign (ex UDT related cruft) or some things that have to be copy/pasted. The heaviest class seems to be the SpanStore variant. Is that a fair assessment?

**Comment by michaelsembwever** [comment]
Yes it is the same schema as far as writes go. Just that the SASI are not used. DSE Solr Cores replace them.

**Comment by codefromthecrypt** [comment]
> Personally I would rather not include the code into master, for the same
> reasons you describe @adriancole https://github.com/adriancole
> 
> But I did want to at least throw the code out here, so that others could
> take it (if only the idea) and run if they wanted to.
> 
> Historically, most of the repos we spun out from concerns about interest
> didn't turn into anything. Ex mongo and redis went to separate repos to
> die. However, I think it is more reasonable that this could live. I'd offer
> effort to carve out a repo for this.

We could attract attention to it, saying that once cassandra3 is done,
we'll deprecate "cassandra" for removal at some point. Those who can't
upgrade on account of DSE should look to the zipkin-storage-dse5 repo. If
no-one is interested enough even to comment there, then there's not much
else we can do... worst case, it dies.

**Comment by codefromthecrypt** [comment]
Thanks for answering my questions, @michaelsembwever! I'll shut up and see what others think

**Comment by schrepfler** [comment]
I would go with external repo and some sort of contrib model and zipkin-storage-dse5 sounds fine (ideally Datastax themselves would be maintainers as they know their features best and it would advocate optimal use of their product.)
Is the DSE Solr interface so much different when starting Cassandra with the parameters to turn on string indexing using ES or Solr (I was under the impression Cassandra ships with such an example config but I can't find it in my instance, perhaps I've seen it in use on some modified Cassandra ex. Titan+Cassandra+ES)?

**Comment by michaelsembwever** [comment]
> Is the DSE Solr interface so much different when starting Cassandra with the parameters to turn on string indexing using ES or Solr …

@schrepfler i'm only familiar with the rather basic approach of using `dsetool create_core …` from the command line to create the Solr Core off a C\* table.

**Comment by schrepfler** [comment]
I think that's the way to go too, haven't dealt with it directly though.

**Comment by danchia** [comment]
AFAIK there's the command line client, and a deprecrated? HTTP endpoint. Sadly can't create the indexes from CQL directly since you need the solr config.

File: zipkin-storage/dse5/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+### Time-To_live
+Time-To-Live is default now at the table level. It can not be overridden in write requests.
+
+### Compaction
+Time-series data is compacted using TimeWindowCompactionStrategy, a known improved over DateTieredCompactionStrategy. Data is
+optimised for queries with a single day. The penalty of reading multiple days is small, a few disk seeks, compared to the
+otherwise overhead of reading a significantly larger amount of data.
+
+### Benchmarking
+Benchmarking the new datamodel demonstrates a significant performance improvement on reads. How much of this translates to te
```
Comment (danchia) [inline zipkin-storage/dse5/README.md:49]: there's a typo here. `translates to the`

Comment (danchia) [reply zipkin-storage/dse5/README.md:49]: Oops sorry, hit finish review too early. Meant to finish my pass later..

File: zipkin-storage/dse5/src/main/java/zipkin/storage/cassandra3/Dse5Schema.java [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+      this.ts = annotation.timestamp;
+      this.v = annotation.value;
+      this.ep = annotation.endpoint != null ? new EndpointUDT(annotation.endpoint) : null;
+    }
+
+    public long getTs() {
+      return ts;
+    }
+
+    public String getV() {
```
Comment (danchia) [inline zipkin-storage/dse5/src/main/java/zipkin/storage/cassandra3/Dse5Schema.java:208]: Minor readability suggestion (@adriancole probably can comment more) but I prefer to not have abbreviations that are not generally well known in public APIs. e.g `ep` probably don't mean much to most people

File: zipkin-storage/dse5/src/main/resources/dse5-schema.cql [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+);
+
+CREATE TYPE IF NOT EXISTS zipkin3.annotation (
+    ts bigint,
+    v  text,
+    ep frozen<endpoint>
+);
+
+CREATE TYPE IF NOT EXISTS zipkin3.binary_annotation (
+    k  text,
```
Comment (danchia) [inline zipkin-storage/dse5/src/main/resources/dse5-schema.cql:17]: I don't recall if in Cassandra 3.x whether long column names inflates storage use, but I would encourage not using abbreviations.

**Review by danchia — COMMENTED** [review]
Maybe I missed it, but perhaps include the solr schema and / or the dsetool command to create the solr core?

**Pull request closed without merging**